posts/2023-03-30-file-expiry-shortcut.md

```diff
@@ -0,0 +1,71 @@
+```
+while read uuid a b c d path; do echo $uuid $path; done
+```
+`list-quarantined` output has 6 columns, but `batch-recover` is expecting a file with 2 columns, which correspond to the first and last columns of the `list-quarantined` output. This loop reads in the `grep`'d `list-quarantined` output line-by-line, saves each column into a different variable, and `echo`'s the ones we need. The rest are discarded. So running those commands connected with pipes gives us this:
+```
+$ nci-file-expiry list-quarantined | grep .ice_daily.nc | while read uuid a b c d path; do echo $uuid $path; done
+813ff7b2-381b-430b-a423-f32b449bf710 /scratch/v45/dr4292/20200222.ice_daily.nc
+9aa317e3-7763-4af5-a19b-d07a7d6c2d90 /scratch/v45/dr4292/20200223.ice_daily.nc
+28b0d00c-e9fc-4e5e-adae-39a817d0fb51 /scratch/v45/dr4292/20200224.ice_daily.nc
+8a17868a-25ec-40bf-a41d-264eb3505da8 /scratch/v45/dr4292/20200225.ice_daily.nc
+...
+```
+```{note}
+There are many ways to orgainse columnated data in `bash`. The `while read echo` variant above is my preference. If you prefer piping to `awk` or `cut`, substitute that instead. As long as the output looks like the output above, the next bit will work.
+```
+To turn this into a "file" that `batch-recover` is happy to deal with, we can take advantage of [process substitution](https://en.wikipedia.org/wiki/Process_substitution). This tricks `batch-recover` into treating the output of the above command it as if it were a file, even though nothing is ever written to disk. So by wrapping the above command in `<( ... )`, its output becomes, for all intents and purposes, the contents of a file.
+
+Process substitution enables a few neat tricks. For instance, if you need to `diff` the output of two commands, you do not need to write the output to temporary files first, you can simply run:
+```
+$ diff <( command_1 ) <( command_2 )
+```
+You can also add an additional `<` operator to redirect this output to `stdin`. The original version of this command posted on slack and the ACCESS-Hive used a this to get the output of `list-quarantine` into the loop.
+```
+while read uuid a b c d path; do echo $uuid $path; done < <( nci-file-expiry list-quarantined | grep .ice_daily.nc )
+```
+There are restrictions on using process substitution, however. The `seek` instruction cannot be used on these "files", meaning that they can't be used in place of structured data (e.g. netCDF). For simple things like this though, that isn't relevant.
+
+This is a lot to remember, so we recommend placing the following in your `~/.bashrc` file:
+```
+function recover_pattern () {
+    nci-file-expiry batch-recover <( nci-file-expiry list-quarantined | grep "${1}" | while read uuid a b c d path; do echo $uuid $path; done )
+}
+```
+And when you log into Gadi again you'll be able to run:
+```
+$ recover_pattern .ice_daily.nc
+```
+for the same effect.
```

Comment:
Is it worth pointing out the limits on what can be in the string passed to the function? Like no regex for example.

I also wonder if there is some utility in making it easy to say "I want to recover files from the directory that I am in". I can imagine users going to a place, finding files that are missing and want to just get those ones back.

Perhaps a function to find missing files from the current directory would also be useful?

Reply:
It might be for the future, but for the moment, considering how busy Dale is right now, I think we should just go ahead as it is. I like the idea of having more tools to make this easier, but I also think we should focus on users managing their storage better and it should be NCI tasks to improve their own file-expiry scripts. You could suggest that to them in your next meeting :-)

Reply:
You can absolutely pass regex to the function. Its just the argument to `grep`, regex is totally valid. As is passing `$PWD`, which will recover everything from the current directory. I'll add a sentence about this in the post.